File: consensus/importFile_align_bis_R2DGC.R
# library(R2DGC)

# result <- ImportFile("D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/A-F-028-817822-droite-ReCIV.txt")
# ImportFile <- function(File) { 
#   MissingStandards <- c()
#   currentRawFile <- read.table(File, sep = "\t", fill = T,  #pour .txt
#                                quote = "", strip.white = T, stringsAsFactors = F, 
#                                header = T)
#   currentRawFile[, 5] <- as.character(currentRawFile[, 
#                                                      5])
#   currentRawFile[, 2] <- as.character(currentRawFile[, 
#                                                      2])
# #filtrage des lignes  avec valeur manquantes
#   currentRawFile <- currentRawFile[which(!is.na(currentRawFile[,3]) & 
#                                         nchar(currentRawFile[,5]) != 0),]


#     # DIAGNOSTIC : Regarder quelques exemples de spectres bruts
#   cat("=== DIAGNOSTIC ===\n")
#   cat("Exemples de spectres bruts (colonne 5):\n")
#   for(i in 1:min(3, nrow(currentRawFile))) {
#     cat("Ligne", i, ":", substr(currentRawFile[i, 5], 1, 100), "...\n")
#   }

#   RTSplit <- data.frame(strsplit(currentRawFile[, 2], " , "), 
#                         stringsAsFactors = F)
#   RTSplit[1, ] <- gsub("\"", "", RTSplit[1, ])
#   RTSplit[2, ] <- gsub("\"", "", RTSplit[2, ])
#   currentRawFile[, "RT1"] <- as.numeric(t(RTSplit[1, ]))
#   currentRawFile[, "RT2"] <- as.numeric(t(RTSplit[2, ]))

#   uniqueIndex <- data.frame(paste(currentRawFile[, 1], 
#                                   currentRawFile[, 2],
#                                   currentRawFile[, 3]))
#   currentRawFile <- currentRawFile[which(!duplicated(uniqueIndex)), 
#   ]
#   row.names(currentRawFile) <- c(1:nrow(currentRawFile))

#    # Parsing des spectres avec diagnostic


#   currentRawFileSplit <- split(currentRawFile, 1:nrow(currentRawFile))
#   spectraSplit <- lapply(currentRawFileSplit, function(a) strsplit(a[[5]], " "))
#   cat("\nAprès split par espaces (premier spectre, premiers éléments):\n")
#   print(spectraSplit[[1]][[1]][1:min(5, length(spectraSplit[[1]][[1]]))])

#   spectraSplit <- lapply(spectraSplit, function(b) lapply(b, function(c) strsplit(c, ":")))
#   cat("\nAprès split par ':' (premier spectre, premiers éléments):\n")
#   print(spectraSplit[[1]][[1]][1:min(3, length(spectraSplit[[1]][[1]]))])

#   spectraSplit <- lapply(spectraSplit, function(d) t(matrix(unlist(d), nrow = 2)))
#    # Vérifier si des éléments sont NULL
#   null_elements <- which(sapply(spectraSplit, is.null))
#   if(length(null_elements) > 0) {
#     cat("Éléments NULL détectés aux positions:", null_elements, "\n")
#   }
#   # Filtrer les NULL
#   spectraSplit <- spectraSplit[!sapply(spectraSplit, is.null)]
#   if(length(spectraSplit) > 0) {
#     cat("\nDimensions de la première matrice:", dim(spectraSplit[[1]]), "\n")
#     cat("Premiers éléments de la première matrice:\n")
#     print(spectraSplit[[1]][1:min(5, nrow(spectraSplit[[1]])), ])

#     spectraSplit <- lapply(spectraSplit, function(d) apply(d, 2, as.numeric))
#     print(spectraSplit[[1]][,1])
  
#     # Filtrer à nouveau les NULL après conversion numérique
#     spectraSplit <- spectraSplit[!sapply(spectraSplit, is.null)]
#     if(length(spectraSplit) > 0) {
#       cat("\nAprès conversion numérique (premiers m/z du premier spectre):\n")
#       print(spectraSplit[[1]][1:min(10, nrow(spectraSplit[[1]])), 1])
#       # Statistiques sur les m/z du premier spectre
#       mz_values <- spectraSplit[[1]][, 1]
#       cat("\nStatistiques m/z premier spectre:\n")
#       cat("Min:", min(mz_values, na.rm = T), "\n")
#       cat("Max:", max(mz_values, na.rm = T), "\n")
#       cat("Nombre de valeurs:", length(mz_values), "\n")
#       cat("Nombre de NA:", sum(is.na(mz_values)), "\n")

#       ionNames <- spectraSplit[[1]][order(spectraSplit[[1]][, 1]), 1]
#       cat("\nPremiers ionNames après tri:\n")
#       print(ionNames[1:min(20, length(ionNames))])
#       cat("\nDerniers ionNames après tri:\n")
#       print(ionNames[max(1, length(ionNames)-19):length(ionNames)])
#   # options(scipen = 999)        
                                              
#       print(ionNames)
#       spectraSplit <- lapply(spectraSplit, function(d) d[order(d[, 1]), 2, drop = F])
#     }
#   }
#   # ionNames <- sort(unique(unlist(lapply(spectraSplit, function(d) d[, 1]))))
                                                       
#   # spectraFull <- spectraSplit                                       
#   # ionNames <- sort(unique(unlist(lapply(spectraFull, function(d) d[,1]))))                                                           
#   # spectraSplit <- lapply(spectraFull, function(d) d[order(d[, 
#   #                                                            1]), 2, drop = F])
#   # spectraSplit <- lapply(spectraSplit, function(m) t(m)) 

#   # spectraFull <- spectraSplit
#   # ionNames <- sort(unique(unlist(lapply(spectraFull, function(d) d[, 1]))))

#   # # Maintenant on garde que les intensités (triées par m/z)
#   # spectraSplit <- lapply(spectraFull, function(d) d[order(d[, 1]), 2, drop = FALSE])

#   return(list(currentRawFile, spectraSplit, MissingStandards, 
#               ionNames, spectraSplit))
#   }

# # Version avec diagnostic pour comprendre le problème
# ImportFile <- function(File) {
#   MissingStandards <- c()
  
#   # Lecture du fichier
#   currentRawFile <- read.table(File, sep = "\t", fill = T, 
#                                quote = "", strip.white = T, 
#                                stringsAsFactors = F, header = T)
#   spectres <- currentRawFile[[5]]

# # Boucle sur chaque spectre (chaîne de caractères)
#   for (i in seq_along(spectres)) {
#     elements <- unlist(strsplit(spectres[i], " "))
#     for (el in elements) {
#       parts <- unlist(strsplit(el, ":"))
#       # Vérifie que l'élément a bien deux parties non vides et convertibles en numérique
#       if (length(parts) != 2 || parts[1] == "" || parts[2] == "" ||
#           is.na(as.numeric(parts[1])) || is.na(as.numeric(parts[2]))) {
#         cat("Entrée mal formée dans la ligne", i, ":", el, "\n")
#       }
#     }
# }
  
#   # Conversion des colonnes en caractères
#   currentRawFile[, 5] <- as.character(currentRawFile[, 5])
#   currentRawFile[, 2] <- as.character(currentRawFile[, 2])
  
#   # Filtrage des lignes avec valeurs manquantes
#   currentRawFile <- currentRawFile[which(!is.na(currentRawFile[,3]) & 
#                                         nchar(currentRawFile[,5]) != 0),]
  
#   # # DIAGNOSTIC : Regarder quelques exemples de spectres bruts
#   # cat("=== DIAGNOSTIC ===\n")
#   # cat("Exemples de spectres bruts (colonne 5):\n")
#   # for(i in 1:min(3, nrow(currentRawFile))) {
#   #   cat("Ligne", i, ":", substr(currentRawFile[i, 5], 1, 100), "...\n")
#   # }
  
#   # Parsing des temps de rétention (inchangé)
#   RTSplit <- data.frame(strsplit(currentRawFile[, 2], " , "), 
#                         stringsAsFactors = F)
#   RTSplit[1, ] <- gsub("\"", "", RTSplit[1, ])
#   RTSplit[2, ] <- gsub("\"", "", RTSplit[2, ])
#   currentRawFile[, "RT1"] <- as.numeric(t(RTSplit[1, ]))
#   currentRawFile[, "RT2"] <- as.numeric(t(RTSplit[2, ]))
  
#   # Suppression des doublons (inchangé)
#   uniqueIndex <- data.frame(paste(currentRawFile[, 1],
#                                  currentRawFile[, 2], 
#                                  currentRawFile[, 3]))
#   currentRawFile <- currentRawFile[which(!duplicated(uniqueIndex)), ]
#   row.names(currentRawFile) <- c(1:nrow(currentRawFile))
  
#   # Parsing des spectres avec diagnostic
#   currentRawFileSplit <- split(currentRawFile, 1:nrow(currentRawFile))
  
#   # Étape 1: Split par espaces
#   spectraSplit <- lapply(currentRawFileSplit, function(a) strsplit(a[[5]], " "))
#   # cat("\nAprès split par espaces (premier spectre, premiers éléments):\n")
#   # print(spectraSplit[[1]][[1]][1:min(5, length(spectraSplit[[1]][[1]]))])
  
#   # Étape 2: Split par ":"
#   spectraSplit <- lapply(spectraSplit, function(b) lapply(b, function(c) strsplit(c, ":")))
#   # cat("\nAprès split par ':' (premier spectre, premiers éléments):\n")
#   # print(spectraSplit[[1]][[1]][1:min(3, length(spectraSplit[[1]][[1]]))])
  
#   # Étape 3: Conversion en matrice
#   spectraSplit <- lapply(spectraSplit, function(d) {
#     tryCatch({
#       t(matrix(unlist(d), nrow = 2))
#     }, error = function(e) {
#       cat("Erreur dans la conversion en matrice:", e$message, "\n")
#       return(NULL)
#     })
#   })
  
#   # Vérifier si des éléments sont NULL
#   null_elements <- which(sapply(spectraSplit, is.null))
#   if(length(null_elements) > 0) {
#     cat("Éléments NULL détectés aux positions:", null_elements, "\n")
#   }
  
#   # Filtrer les NULL
#   spectraSplit <- spectraSplit[!sapply(spectraSplit, is.null)]
  
#   if(length(spectraSplit) > 0) {
#     # cat("\nDimensions de la première matrice:", dim(spectraSplit[[1]]), "\n")
#     # cat("Premiers éléments de la première matrice:\n")
#     # print(spectraSplit[[1]][1:min(5, nrow(spectraSplit[[1]])), ])
    
#     # Conversion numérique
#     # spectraSplit <- lapply(spectraSplit, function(d) {
#     #   tryCatch({
#     #     apply(d, 2, as.numeric)
#     #   }, error = function(e) {
#     #     cat("Erreur dans la conversion numérique:", e$message, "\n")
#     #     return(NULL)
#     #   })
#     # })
#     spectraSplit <- lapply(spectraSplit, function(d) {
#     d <- apply(d, 2, as.numeric)
#     if (any(is.na(d[,1]))) {
#       # cat("Spectre avec NA dans m/z:\n")
#       # print(d[is.na(d[,1]), ])
#       d <- d[!is.na(d[,1]), , drop = FALSE]
#     }
#     return(d)
#   })
    
#     # Filtrer à nouveau les NULL après conversion numérique
#     spectraSplit <- spectraSplit[!sapply(spectraSplit, is.null)]
    
#     if(length(spectraSplit) > 0) {
#       cat("\nAprès conversion numérique (premiers m/z du premier spectre):\n")
#       print(spectraSplit[[1]][1:min(10, nrow(spectraSplit[[1]])), 1])
      
#       # Statistiques sur les m/z du premier spectre
#       mz_values <- spectraSplit[[1]][, 1]
#       cat("\nStatistiques m/z premier spectre:\n")
#       cat("Min:", min(mz_values, na.rm = T), "\n")
#       cat("Max:", max(mz_values, na.rm = T), "\n")
#       cat("Nombre de valeurs:", length(mz_values), "\n")
#       cat("Nombre de NA:", sum(is.na(mz_values)), "\n")
#       # Noms des ions (m/z triés)
#       # spectraSplit_clean <- lapply(spectraSplit_num, function(d) d[!is.na(d[,1]) & !is.na(d[,2]), ]) 
#       # spectraSplit_sorted <- lapply(spectraSplit_clean, function(d) d[order(d[, 1]), 2, drop = FALSE])
#       # ionNames <- spectraSplit_clean[[1]][order(spectraSplit_clean[[1]][, 1]), 1]
#       ionNames <- spectraSplit[[1]][order(spectraSplit[[1]][, 1]), 1]
      
#       cat("\nPremiers ionNames après tri:\n")
#       print(ionNames[1:min(20, length(ionNames))])
#       cat("\nDerniers ionNames après tri:\n")
#       print(ionNames[max(1, length(ionNames)-19):length(ionNames)])
      
#       # Réorganisation des spectres par ordre croissant de m/z
#       spectraSplit <- lapply(spectraSplit, function(d) d[order(d[, 1]), 2, drop = F])
      
#       return(list(currentRawFile, spectraSplit, MissingStandards, ionNames))
#     }
#   }
  
#   cat("Erreur: Impossible de parser les spectres\n")
#   return(NULL)
# }

ImportFile <- function(File) { 

  MissingStandards <- c()
  currentRawFile <- read.table(File, sep = "\t", fill = T, 
                               quote = "", strip.white = T, stringsAsFactors = F, 
                               header = T)

  currentRawFile[, 5] <- as.character(currentRawFile[, 
                                                     5])
  currentRawFile[, 2] <- as.character(currentRawFile[, 
                                                     2])
  # #  Filtrage des lignes avec valeurs manquantes
  # currentRawFile <- currentRawFile[which(!is.na(currentRawFile[,3]) & 
  #                                       nchar(currentRawFile[,5]) != 0),]

  RTSplit <- data.frame(strsplit(currentRawFile[, 2], " , "), 
                        stringsAsFactors = F)

  RTSplit[1, ] <- gsub("\"", "", RTSplit[1, ])
  RTSplit[2, ] <- gsub("\"", "", RTSplit[2, ])
  currentRawFile[, "RT1"] <- as.numeric(t(RTSplit[1, ]))
  currentRawFile[, "RT2"] <- as.numeric(t(RTSplit[2, ]))
  uniqueIndex <- data.frame(paste(currentRawFile[, 1], 
                                  currentRawFile[, 2], currentRawFile[, 3]))

  currentRawFile <- currentRawFile[which(!duplicated(uniqueIndex)), 
  ]

  row.names(currentRawFile) <- c(1:nrow(currentRawFile))


  currentRawFileSplit <- split(currentRawFile, 1:nrow(currentRawFile))

  spectraSplit <- lapply(currentRawFileSplit, function(a) strsplit(a[[5]], 
                                                                   " "))

  spectraSplit <- lapply(spectraSplit, function(b) lapply(b, 
                                                          function(c) strsplit(c, ":")))

  spectraSplit <- lapply(spectraSplit, function(d) t(matrix(unlist(d), 
                                                            nrow = 2)))

  

  spectraSplit <- lapply(spectraSplit, function(d) apply(d, 
                                                         2, as.numeric))

  ionNames <- spectraSplit[[1]][order(spectraSplit[[1]][, 
                                                        1]), 1]

  spectraSplit <- lapply(spectraSplit, function(d) d[order(d[, 
                                                             1]), 2, drop = F])

  return(list(currentRawFile, spectraSplit, MissingStandards, 
              ionNames, spectraSplit))

}


fichier <- "D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/A-F-028-817822-droite-ReCIV.txt"
res <- ImportFile(fichier)
to_save <- list(
  main_table = res[[1]],
  spectra_split = res[[2]],
  missing_standards = res[[3]],
  ion_names = res[[4]]
)
str(res)
# str(to_save[["ion_names"]])
ionNames<- res[[1]][, "Spectra"]
head(ionNames)
unique_ions <- sort(unique(as.numeric(ionNames)))
print(unique_ions)
write.csv(res[[1]], "D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/r_current_raw_file.csv", row.names = F)
saveRDS(to_save,file="D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/r_spectra_split.rds")
write.csv(res[[4]], file = "D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/r_ion_names.csv", row.names = FALSE)


# write.csv(data.frame(ion = res[[1]]$Spectra), "D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/r_ion_names.csv", row.names = F)
# ions_char <- sapply(res[[4]], function(x) paste(x, collapse = " "))
# write.csv(data.frame(ion = ions_char), "D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/r_ion_names.csv", row.names = FALSE)

File: archives/consensus_test/importFile_align_bis_R2DGC.py
import pandas as pd
import numpy as np

# def importFile(file):
#     missing_standards = []

#     #read the file    
#     current_raw_file = pd.read_csv(file, sep="\t", header=0,skipinitialspace=True)
#     current_raw_file = current_raw_file.applymap(lambda x: x.strip() if isinstance(x, str) else x)
#     print(current_raw_file.iloc[:, 1].head())

#     # Convert columns to string
#     current_raw_file.iloc[:, 4] = current_raw_file.iloc[:, 4].astype(str)
#     current_raw_file.iloc[:, 1] = current_raw_file.iloc[:, 1].astype(str)

#     # rajoute les ""
#     current_raw_file['R.T...s.'] = '"' + current_raw_file['R.T...s.'].str.strip('"') + '"'
    
#     #split RTs
#     current_raw_file[["RT1", "RT2"]] = current_raw_file["R.T...s."].str.replace('"', '', regex=False).str.split(" , ", expand=True).astype(float)

#     # suppression des doublons
#     composite_key = composite_key = current_raw_file["Name"].astype(str) + "_" + \
#                 current_raw_file["R.T...s."].astype(str) + "_" + \
#                 current_raw_file["Area"].astype(str)
#     current_raw_file = current_raw_file.loc[~composite_key.duplicated()].reset_index(drop=True)

#     # Spectres en liste par ligne (chaque spectre est une liste d'intensités)
#     spectra_split = []
#     ion_names = None
#     for i, row in current_raw_file.iterrows():
#         spectrum = row.iloc[4]
#         peak_list = [p.split(":") for p in spectrum.strip().split(" ") if ":" in p]
#         try:
#             mzs, intensities = zip(*[(float(mz), float(intensity)) for mz, intensity in peak_list])
#         except ValueError:
#             mzs, intensities = [], []

#         # On trie par m/z croissant
#         sorted_pairs = sorted(zip(mzs, intensities), key=lambda x: x[0])
#         sorted_mzs, sorted_intensities = zip(*sorted_pairs) if sorted_pairs else ([], [])
#         if ion_names is None and sorted_mzs:
#             ion_names = list(sorted_mzs)
        
#         spectra_split.append(np.array(sorted_intensities))

#     return [current_raw_file, spectra_split, missing_standards, ion_names, spectra_split]

def importFile(file):
    """Import and process chromatographic data file"""
    missing_standards = []

    #read the file    
    current_raw_file = pd.read_csv(file, sep="\t", header=0,skipinitialspace=True)
    current_raw_file = current_raw_file.apply(lambda col: col.map(lambda x: x.strip() if isinstance(x, str) else x))

    # print(current_raw_file.iloc[:, 1].head())

    # Convert columns to string
    current_raw_file.iloc[:, 4] = current_raw_file.iloc[:, 4].astype(str)
    current_raw_file.iloc[:, 1] = current_raw_file.iloc[:, 1].astype(str)

# rajoute les ""
    # current_raw_file['R.T...s.'] = '"' + current_raw_file['R.T...s.'].str.strip('"') + '"'
    # rajoute les ""
    current_raw_file['R.T...s.'] = '"' + current_raw_file['R.T...s.'].str.strip('"') + '"'
    
    # #split RTs
    # current_raw_file[["RT1", "RT2"]] = current_raw_file["R.T...s."].str.replace('"', '', regex=False).str.split(" , ", expand=True).astype(float)

    # # suppression des doublons
    # composite_key = composite_key = current_raw_file["Name"].astype(str) + "_" + \
    #             current_raw_file["R.T...s."].astype(str) + "_" + \
    #             current_raw_file["Area"].astype(str)
    # current_raw_file = current_raw_file.loc[~composite_key.duplicated()].reset_index(drop=True)

     # Split RTs - remove quotes and split on " , "
    rt_split = current_raw_file.iloc[:, 1].str.replace('"', '', regex=False).str.split(' , ', expand=True)
    current_raw_file["RT1"] = rt_split[0].astype(float)
    current_raw_file["RT2"] = rt_split[1].astype(float)

    # Create unique index for duplicate removal (Name + RT + Area)
    unique_index = (current_raw_file.iloc[:, 0].astype(str) + 
                   current_raw_file.iloc[:, 1].astype(str) + 
                   current_raw_file.iloc[:, 2].astype(str))
    
    # Remove duplicates
    current_raw_file = current_raw_file.loc[~unique_index.duplicated()].reset_index(drop=True)

    # Spectres en liste par ligne (chaque spectre est une liste d'intensités)
    spectra_split = []
    ion_names = None

    # for i, row in current_raw_file.iterrows():
    #     spectrum = row.iloc[4]
    #     peak_list = [p.split(":") for p in spectrum.strip().split(" ") if ":" in p]
    #     try:
    #         mzs, intensities = zip(*[(float(mz), float(intensity)) for mz, intensity in peak_list])
    #     except ValueError:
    #         mzs, intensities = [], []

    #     # On trie par m/z croissant
    #     sorted_pairs = sorted(zip(mzs, intensities), key=lambda x: x[0])
    #     sorted_mzs, sorted_intensities = zip(*sorted_pairs) if sorted_pairs else ([], [])
    #     if ion_names is None and sorted_mzs:
    #         ion_names = list(sorted_mzs)
        
    #     spectra_split.append(np.array(sorted_intensities))
    for i, row in current_raw_file.iterrows():
        spectrum = row.iloc[4]  # Column 5 in R = index 4 in Python
        if pd.isna(spectrum) or spectrum == '':
            spectra_split.append(np.array([]))
            continue
            
        # Split spectrum by spaces, then by colons
        peak_list = []
        for peak in spectrum.strip().split(" "):
            if ":" in peak:
                parts = peak.split(":")
                if len(parts) == 2:
                    try:
                        mz = float(parts[0])
                        intensity = float(parts[1])
                        peak_list.append((mz, intensity))
                    except ValueError:
                        continue
        
        if peak_list:
            # Sort by m/z
            peak_list.sort(key=lambda x: x[0])
            mzs, intensities = zip(*peak_list)
            
            # Set ion_names from first spectrum
            if ion_names is None:
                ion_names = list(mzs)
            
            spectra_split.append(np.array(intensities))
        else:
            spectra_split.append(np.array([]))

    return [current_raw_file, spectra_split, missing_standards, ion_names, spectra_split]

if __name__ == "__main__":

    #DEBUG
    file = "D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/A-F-028-817822-droite-ReCIV.txt"
    with open(file, "r", encoding="utf-8") as f:
        line = f.readlines()
        print(len(line))
        print(line[584])



    # importFile("D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/A-F-028-817822-droite-ReCIV.txt")
    df, spectra_split, missing_standards, ion_names, spectra_split = importFile("D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/A-F-028-817822-droite-ReCIV.txt")
    df.to_csv("D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/py_current_raw_file.csv", index=False)

    import pickle
    with open("D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/py_spectra_split.pkl", "wb") as f:
        pickle.dump(spectra_split, f)   

    pd.DataFrame({"ion": ion_names}).to_csv("D:/Dossiers Persos/Adeline/Python-2DGC-Alignment/consensus/py_ion_names.csv", index=False)